File: solutions/167.two-sum-ii-input-array-is-sorted.js
/*
 * @lc app=leetcode id=167 lang=javascript
 *
 * [167] Two Sum II - Input Array Is Sorted
 */

// @lc code=start
/**
 * @param {number[]} numbers
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (numbers, target) {
  let i = 0,
    j = numbers.length - 1;
  while (i < j) {
    if (numbers[i] + numbers[j] < target) {
      ++i;
    } else if (numbers[i] + numbers[j] > target) {
      --j;
    } else {
      break;
    }
  }
  return [i + 1, j + 1];
};
// @lc code=end

File: solutions/167.two-sum-ii-input-array-is-sorted.rb
# @param {Integer[]} numbers
# @param {Integer} target
# @return {Integer[]}
def two_sum(numbers, target)
  i = 0
  j = numbers.size - 1
  while i <= j
    sum = numbers[i] + numbers[j]
    if sum < target
      i += 1
    elsif sum > target
      j -= 1
    else
      return [i + 1, j + 1]
    end
  end
  -1
end

# 双指针，缩减搜索空间（虽然看起来像BS - O(logN)，但完全不同 - O(N) ）
